src/server.js

```diff
@@ -1,168 +1,174 @@
+const chromiumExec = process.env.SCREENIE_CHROMIUM_EXEC
+  ? { executablePath: process.env.SCREENIE_CHROMIUM_EXEC }
+  : {};
 
-/*
- * Clean up the PhantomJS pool before exiting when receiving a termination
- * signal. Exit with status code 143 (128 + SIGTERM's signal number, 15).
- */
-process.on('SIGTERM', () => {
-    logger.info('Received SIGTERM, exiting...');
-    pool.drain()
-        .then(() => pool.clear())
-        .then(() => process.exit(143));
-});
+let browser;
 
 /**
- * Set up a PhantomJS instance with a page and configure viewport size.
+ * Set up a Puppeteer instance with a page and configure viewport size.
  */
-app.use(function *(next) {
-    const size = {
-        width: Math.min(
-            2048,
-            parseInt(this.request.query.width, 10) || imageSize.width
-        ),
-        height: Math.min(
-            2048,
-            parseInt(this.request.query.height, 10) || imageSize.height
-        ),
-    };
-
-    logger.verbose(
-        `Instantiating PhantomJS with page size ${size.width}x${size.height}`
+app.use(function*(next) {
+  const { width, height } = this.request.query;
+  const size = {
+    width: Math.min(2048, parseInt(width, 10) || imageSize.width),
+    height: Math.min(2048, parseInt(height, 10) || imageSize.height),
+  };
+  let pageError;
+
+  logger.verbose(
+    `Instantiating Puppeteer with page size ${size.width}x${size.height}`
+  );
+
+  if (!browser) {
+    yield puppeteer
+      .launch({ ...chromiumExec, ...chromiumArgs })
+      .then(instance => (browser = instance));
+  }
+
+  yield browser
+    .newPage()
+    .then(page => (this.state.page = page))
+    .then(() => this.state.page.setViewport(size))
+    .catch(error => (pageError = error));
+
+  if (pageError) {
+    this.throw(
+      400,
+      'Could not open a page due to an error: ' + pageError.message
     );
+  }
 
-    yield pool.use(instance => instance.createPage())
-        .then(page => this.state.page = page)
-        .then(() => this.state.page.property('viewportSize', size));
-
-    yield next;
+  yield next;
 });
 
 /**
- * Attempt to load given URL in the PhantomJS page.
+ * Attempt to load given URL in the Puppeteer page.
  *
  * Throws 400 Bad Request if no URL is provided, and 404 Not Found if
- * PhantomJS could not load the URL.
+ * Puppeteer could not load the URL.
  */
-app.use(function *(next) {
-    const { url } = this.request.query;
+app.use(function*(next) {
+  const { url } = this.request.query;
+  const { page } = this.state;
 
-    if (!url) {
-        this.throw('No url request parameter supplied.', 400);
-    }
+  if (!url) {
+    this.throw(400, 'No url request parameter supplied.');
+  }
 
-    logger.verbose(`Attempting to load ${url}`);
+  logger.verbose(`Attempting to load ${url}`);
 
-    yield this.state.page.open(url)
-        .then(status => status === 'success')
-        .then(loaded => loaded || this.throw(404));
+  yield page.goto(url).catch(error => {
+    this.throw(404);
```

Comment:
Isn't this susceptible to the `UnhandledPromiseRejection` issue as well, since we're throwing inside the catch handler?